### PR TITLE
[SPARK-20680][SQL] Adding HiveVoidType in Spark to be compatible with Hive

### PR DESCRIPTION
**Pull request by LantaoJin**
### What changes were proposed in this pull request?

This is to address the close one #17953, as a refactor for #28833 
Adding `HiveVoidType`, like `HiveStringType`, to prevent from exception when describe tables/views which the schema contain Hive VOID/NULL type.

### Why are the changes needed?

Spark is incompatible with hive void type. When Hive table schema contains void type, DESC table will throw an exception in Spark.

>hive> create table bad as select 1 x, null z from dual;
>hive> describe bad;
OK
x	int	
z	void

In Spark2.0.x, the behaviour to read this view is normal:
>spark-sql> describe bad;
x       int     NULL
z       void    NULL
Time taken: 4.431 seconds, Fetched 2 row(s)

But in lastest Spark version, it failed with SparkException: Cannot recognize hive type string: void

>spark-sql> describe bad;
17/05/09 03:12:08 ERROR thriftserver.SparkSQLDriver: Failed in [describe bad]
org.apache.spark.SparkException: Cannot recognize hive type string: void
Caused by: org.apache.spark.sql.catalyst.parser.ParseException:
DataType void() is not supported.(line 1, pos 0)
== SQL ==  
void       
^^^
        ... 61 more
org.apache.spark.SparkException: Cannot recognize hive type string: void

### Does this PR introduce any user-facing change?

No

### How was this patch tested?

Add unit tests

Also can manual tests
>spark-sql> describe bad;
x int NULL
z null NULL
Time taken: 0.486 seconds, Fetched 2 row(s)

### CODE REVIEW

**Review by HyukjinKwon — COMMENTED** [review]
We should also fix https://github.com/apache/spark/blob/68d7edf9497bea2f73707d32ab55dd8e53088e7c/python/pyspark/sql/tests/test_types.py#L486 and add `null` type at https://github.com/apache/spark/blob/d21aab403a0a32e8b705b38874c0b335e703bd5d/python/pyspark/sql/types.py#L752

I am okay if you're not used to Python side - I can do it in a followup.

**Comment by LantaoJin** [comment]
Thanks @HyukjinKwon, if this could be merged, can you help on python side?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 1]
```diff
@@ -2184,7 +2184,9 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
    * Create a Spark DataType.
    */
   private def visitSparkDataType(ctx: DataTypeContext): DataType = {
-    HiveStringType.replaceCharType(typedVisit(ctx))
+    HiveNullType.replaceNullType(
+      HiveStringType.replaceCharType(typedVisit(ctx))
+    )
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2189]: One line is enough, isn't it?
```scala
HiveNullType.replaceNullType(HiveStringType.replaceCharType(typedVisit(ctx)))
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 2]
```diff
@@ -2260,7 +2263,9 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
 
     // Add Hive type string to metadata.
     val rawDataType = typedVisit[DataType](ctx.dataType)
-    val cleanedDataType = HiveStringType.replaceCharType(rawDataType)
+    val cleanedDataType = HiveNullType.replaceNullType(
+      HiveStringType.replaceCharType(rawDataType)
+    )
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2268]: ditto.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 1 above)
Comment (LantaoJin) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2189]: Yes. Split to two lines is just for readability. Ok, I will coalesce to one line.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/HiveNullType.scala [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+ */
+
+package org.apache.spark.sql.types
+
+/**
+ * A hive null type for compatibility. These datatypes should only used for parsing,
+ * and should NOT be used anywhere else. Any instance of these data types should be
+ * replaced by a [[NullType]] before analysis.
+ */
+class HiveNullType private() extends DataType {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/HiveNullType.scala:25]: I know the context, but can we name this `HiveVoidType` literally?

Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/HiveNullType.scala:25]: Currently, the description is interpreted like "`hive null type` should be replaced by a NullType before analysis".

Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/HiveNullType.scala:25]: `null` is a `value` and Hive exposes `void` as a type.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/HiveNullType.scala [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+  override def defaultSize: Int = 1
+
+  override private[spark] def asNullable: HiveNullType = this
+
+  override def simpleString: String = "void"
+}
+
+case object HiveNullType extends HiveNullType {
+  def replaceNullType(dt: DataType): DataType = dt match {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/HiveNullType.scala:35]: This will be `replaceVoidType` and it will be less ambiguous.

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala [hunk 5]
```diff
@@ -132,7 +132,9 @@ class ResolveSessionCatalog(
               }
           }
           // Add Hive type string to metadata.
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:134]: Please update this description together in this PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/HiveNullType.scala (hunk 3 above)
Comment (LantaoJin) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/HiveNullType.scala:25]: > `null` is a `value` and Hive exposes `void` as a type.

You are right.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/HiveNullType.scala [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    case ArrayType(et, nullable) =>
+      ArrayType(replaceNullType(et), nullable)
+    case MapType(kt, vt, nullable) =>
+      MapType(replaceNullType(kt), replaceNullType(vt), nullable)
+    case StructType(fields) =>
+      StructType(fields.map { field =>
+        field.copy(dataType = replaceNullType(field.dataType))
+      })
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/HiveNullType.scala:43]: Maybe, the following is shorter as a one-liner.
```scala
StructType(fields.map(f => f.copy(dataType = replaceNullType(f.dataType))))
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/HiveNullType.scala [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    case _: HiveNullType => NullType
+    case _ => dt
+  }
+
+
+  def containsNullType(dt: DataType): Boolean = dt match {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/HiveNullType.scala:49]: Shall we remove this because this PR doesn't use this at all? You can add later when you need.

**Comment by dongjoon-hyun** [comment]
Thank you for working on this, @LantaoJin !

**Comment by LantaoJin** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
Thank you for updating, @LantaoJin .

**Comment by dongjoon-hyun** [comment]
@LantaoJin . Is there a reason why you use `$` at the Scala file name?
- sql/catalyst/src/main/scala/org/apache/spark/sql/types/HiveVoidType$.scala

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 8]
```diff
@@ -2309,6 +2309,22 @@ class HiveDDLSuite
     }
   }
 
+  test("SPARK-20680: Spark-sql do not support for void column datatype of view") {
+    withTable("t") {
+      withView("tabNullType") {
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala:2314]: `tblNullType` -> `tblVoidType`.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 9]
```diff
@@ -2309,6 +2309,22 @@ class HiveDDLSuite
...
+        val client =
+          spark.sharedState.externalCatalog.unwrapped.asInstanceOf[HiveExternalCatalog].client
+        client.runSqlHive("CREATE TABLE t (t1 int)")
+        client.runSqlHive("INSERT INTO t VALUES (3)")
+        client.runSqlHive("CREATE VIEW tabNullType AS SELECT NULL AS col FROM t")
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala:2319]: `tabNullType` -> `tabVoidType`

**Comment by LantaoJin** [comment]
> @LantaoJin . Is there a reason why you use `$` at the Scala file name?
> 
> * sql/catalyst/src/main/scala/org/apache/spark/sql/types/HiveVoidType$.scala

No, just typo. Fixed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/HiveStringType.scala [hunk 10]
```diff
@@ -47,9 +47,7 @@ object HiveStringType {
     case MapType(kt, vt, nullable) =>
       MapType(replaceCharType(kt), replaceCharType(vt), nullable)
     case StructType(fields) =>
-      StructType(fields.map { field =>
-        field.copy(dataType = replaceCharType(field.dataType))
-      })
+      StructType(fields.map(f => f.copy(dataType = replaceCharType(f.dataType))))
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/HiveStringType.scala:50]: Ur, is this relevant to this `HiveVoidType` PR?

Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/HiveStringType.scala:50]: @LantaoJin . I understand that you want to match the style, but a PR should be minimal and should not touch irrelevant parts.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 11]
```diff
@@ -2212,6 +2212,7 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
       case ("decimal" | "dec" | "numeric", precision :: scale :: Nil) =>
         DecimalType(precision.getText.toInt, scale.getText.toInt)
       case ("interval", Nil) => CalendarIntervalType
+      case ("void", Nil) => HiveVoidType
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2215]: We can just reuse `NullType`. We should forbid creating table with null type completely, including `spark.catalog.createTable`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/HiveStringType.scala (hunk 10 above)
Comment (LantaoJin) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/HiveStringType.scala:50]: Ok, get the rule now. I will rollback this line

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 11 above)
Comment (LantaoJin) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2215]: Em, yes. We don't need this line in this PR, right?

**Comment by cloud-fan** [comment]
I think we need two changes:
1. forbid creating tables with void column type. This could be done in the rule `CommandCheck`. We can test it with `spark.catalog.createTable`
2. support "void" in the parser to fix SPARK-20680. This is only for legacy hive tables, as new tables can't have void type columns.

These two changes can be done with 2 PRs.

**Comment by LantaoJin** [comment]
Yes. This PR is for the second change.

**Comment by cloud-fan** [comment]
we need to do 1 first, otherwise it makes users be able to create tables with void type via CREATE TABLE command, while it was not possible before as the parser doesn't support it.

**Comment by LantaoJin** [comment]
Ah, now I understood the context. The 5th commit should be reverted in this PR and we need to do 1 first. I will work on that tomorrow. My laptop is out of power now.

**Comment by LantaoJin** [comment]
@cloud-fan I refactor some codes, now I think this PR could be no dependency.

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala [hunk 12]
```diff
@@ -131,8 +135,9 @@ class ResolveSessionCatalog(
                     s"Available: ${v1Table.schema.fieldNames.mkString(", ")}")
               }
           }
-          // Add Hive type string to metadata.
-          val cleanedDataType = HiveStringType.replaceCharType(dataType)
+          // Add Hive type 'string' and 'void' to metadata.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:138]: we can be more aggressive here: forbid void type in all cases, including hive tables.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 13]
```diff
@@ -2184,7 +2184,7 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
    * Create a Spark DataType.
    */
   private def visitSparkDataType(ctx: DataTypeContext): DataType = {
-    HiveStringType.replaceCharType(typedVisit(ctx))
+    HiveVoidType.replaceVoidType(HiveStringType.replaceCharType(typedVisit(ctx)))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2187]: I don't get it why we need `HiveVoidType`. What happens if we just parse `void` to `NullType`?

Comment (LantaoJin) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2187]: Because that could indicate VOID is a Hive type, the handle processing is more unified. Or, we can just use the PR #28833

Comment (LantaoJin) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2187]: For example,  below function will point the failure is due to the legacy hive void type. If we mix VOID and NULL, I am not sure it would be better than separation.
```scala
  def failVoidType(dt: DataType): Unit = {
    if (HiveVoidType.containsVoidType(dt)) {
      throw new AnalysisException(
        "Cannot create tables with Hive VOID type.")
    }
  }
```

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2187]: VOID and NULL are indeed the same type. We can just check null type and fail with error message: `Cannot create tables with VOID type`

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2187]: The point is consistency: The VOID type in SQL statement should be the same as `NullType` specified by Scala API in `spark.catalog.createTable`.

Comment (LantaoJin) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2187]: @cloud-fan , ok. I will follow your suggestion to fix it in #28833 , since this PR is a refator with new type `HiveVoidType`. Now we don't need it.

**Comment by LantaoJin** [comment]
Close this since #28833 merged. Thank you!
